### PR TITLE
storage: ensure no filesystem use after close in test builds

### LINKED ISSUE #110293: kvserver: storage engine is used after being closed in TestCheckConsistencyInconsistent
Pick up #106195, and then we get this (found when working on #106459)
```
=== RUN   TestCheckConsistencyInconsistent
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/59d2a1c5950246cd61480297d44235f9/logTestCheckConsistencyInconsistent2374593241
    test_log_scope.go:81: use -show-logs to present logs inline
    consistency_queue_test.go:382: found a checkpoint at auxiliary/checkpoints/r5_at_30
    consistency_queue_test.go:382: found a checkpoint at auxiliary/checkpoints/r5_at_30
    consistency_queue_test.go:382: found a checkpoint at auxiliary/checkpoints/r5_at_30
*
* ERROR: a panic has occurred!
* engine is closed; the Engine-provided filesystem cannot be used after Close
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:521
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:890
*   | [...repeated from below...]
* Wraps: (2) assertion failure
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/storage.(*noUseAfterClose).ensureStillOpen
*   | 	github.com/cockroachdb/cockroach/pkg/storage/pebble.go:2787
*   | github.com/cockroachdb/cockroach/pkg/storage.(*noUseAfterClose).Open
*   | 	github.com/cockroachdb/cockroach/pkg/storage/pebble.go:2856
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestCheckConsistencyInconsistent
*   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/consistency_queue_test.go:416
*   | testing.tRunner
*   | 	GOROOT/src/testing/testing.go:1446
*   | runtime.goexit
*   | 	GOROOT/src/runtime/asm_amd64.s:1594
* Wraps: (4) engine is closed; the Engine-provided filesystem cannot be used after Close
* Error types: (1) *withstack.withStack (2) *assert.withAssertionFailure (3) *withstack.withStack (4) *errutil.leafError
*
    stopper.go:230: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: /artifacts/tmp/_tmp/59d2a1c5950246cd61480297d44235f9/logTestCheckConsistencyInconsistent2374593241
    stopper.go:230: panic: engine is closed; the Engine-provided filesystem cannot be used after Close
--- FAIL: TestCheckConsistencyInconsistent (11.31s)
panic: engine is closed; the Engine-provided filesystem cannot be used after Close [recovered]
	panic: engine is closed; the Engine-provided filesystem cannot be used after Close [recovered]
	panic: engine is closed; the Engine-provided filesystem cannot be used after Close [recovered]
	panic: engine is closed; the Engine-provided filesystem cannot be used after Close [recovered]
	panic: engine is closed; the Engine-provided filesystem cannot be used after Close

goroutine 19134 [running]:
testing.tRunner.func1.2({0x60c3280, 0xc004e7e9e0})
	GOROOT/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1399 +0x39f
panic({0x60c3280, 0xc004e7e9e0})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:133 +0x2f5
panic({0x60c3280, 0xc004e7e9e0})
	GOROOT/src/runtime/panic.go:890 +0x262
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x0?, {0x7f9a090, 0xc0000c0070})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x6a
panic({0x60c3280, 0xc004e7e9e0})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc00abcb0e0, {0x7f9a090, 0xc0000c0070})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:521 +0x19b
panic({0x60c3280, 0xc004e7e9e0})
	GOROOT/src/runtime/panic.go:890 +0x262
github.com/cockroachdb/cockroach/pkg/storage.(*noUseAfterClose).ensureStillOpen(...)
	github.com/cockroachdb/cockroach/pkg/storage/pebble.go:2787
github.com/cockroachdb/cockroach/pkg/storage.(*noUseAfterClose).Open(0x56200a?, {0xc008fa4ba0?, 0x1d?}, {0x0?, 0x7eefbfc?, 0xc0069c2c01?})
	github.com/cockroachdb/cockroach/pkg/storage/pebble.go:2856 +0x116
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestCheckConsistencyInconsistent(0xc0053ea1a0)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/consistency_queue_test.go:416 +0x165f
testing.tRunner(0xc0053ea1a0, 0x6b2a7c0)
	GOROOT/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
	GOROOT/src/testing/testing.go:1493 +0x35f
```

Jira issue: CRDB-31358

### PR DESCRIPTION
**Pull request by jbowens**
In test builds, panic if any of the VFS filesystem facilities are used after the Engine is closed.

Epic: none
Informs #106119.
Resolves #110293.
Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
Hey @jbowens, should we push this over the finish line? I ran only into a single failure #110293 when I polished #106459 last week (with this PR cherry-picked), so I think it should be almost good to go.

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

Nice!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106195#-:-Ne9sRCB9-J5_x8dMUqM:bli1pub)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal)

**Comment by erikgrinaker** [comment]
Nice, thanks for the test fix! We hadn't gotten around to this yet.

File: pkg/kv/kvserver/consistency_queue_test.go [hunk 1]
```diff
@@ -337,6 +337,7 @@ func TestCheckConsistencyInconsistent(t *testing.T) {
 
 	// Write some arbitrary data only to store on n2. Inconsistent key "e"!
 	s2 := tc.GetFirstStoreFromServer(t, 1)
+	s2AuxDir := s2.TODOEngine().GetAuxiliaryDir()
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/consistency_queue_test.go:340]: nit: comment why we can't just inline this below, e.g.:

```golang
// Remember the path to s2 aux directory in storage. We can't use Engine
// after s2 crashes below, so will read from its filesystem directly.
```

File: pkg/kv/kvserver/consistency_queue_test.go [hunk 2]
```diff
@@ -412,8 +413,8 @@ func TestCheckConsistencyInconsistent(t *testing.T) {
 	assert.NotEqual(t, hashes[0], hashes[1]) // s2 diverged
 
 	// A death rattle should have been written on s2.
-	eng := s2.TODOEngine()
-	f, err := eng.Open(base.PreventedStartupFile(eng.GetAuxiliaryDir()))
+	fs := stickyVFSRegistry.Get(strconv.FormatInt(int64(1), 10))
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/consistency_queue_test.go:416]: nit: could just inline

```golang
fs := stickyVFSRegistry.Get("1")
```

**Review by pav-kv — COMMENTED** [review]
The consistency checker test fix LGTM % nits. Thanks for fixing!

**Review by pav-kv — APPROVED** [review]

**Review by jbowens — COMMENTED** [review]
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106195#-:-NeJbrBI4AzAvP_nBkUG:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/consistency_queue_test.go` line 340 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106195#gh-1326217652:-NeJalX_7AINJdhk1KaB:b-bicsgv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8deacf6d320ae24d8d3dca425210342460436bbf/pkg/kv/kvserver/consistency_queue_test.go#L340)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

nit: comment why we can't just inline this below, e.g.:

```golang
// Remember the path to s2 aux directory in storage. We can't use Engine
// after s2 crashes below, so will read from its filesystem directly.
```
</blockquote></details>

Not sure I follow, but the Store and Engine are stopped down below where we use `s2AuxDir`. 
___
*[`pkg/kv/kvserver/consistency_queue_test.go` line 416 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106195#gh-1326223818:-NeJbk6-DyXepufFK4d0:b2juzuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8deacf6d320ae24d8d3dca425210342460436bbf/pkg/kv/kvserver/consistency_queue_test.go#L416)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

nit: could just inline

```golang
fs := stickyVFSRegistry.Get("1")
```
</blockquote></details>

Heh, was mindlessly copy-pasting. Done.

**Comment by jbowens** [comment]
TFTRs!

bors r+

**Pull request merged**